man/emu-verify.1
.TH emu-verify 1 July 14, 2015 "version 0.2.0" "Emu Manual"
.SH NAME
emu\-verify \- Verify the contents of snapshots
.SH SYNOPSIS
.B emu verify
[<options>] [<sink>[:<snapshot>] ...]
.SH DESCRIPTION
Verifies snapshot checksums by computing against the stored tree. A
checksum which does not match is indicative of a snapshot that is in
some way corrupted or has been tampered with.
.PP
The format for specifying snapshots is in the form
<sink>[:<snapshot>]. The <snapshot> value is the 40 digit snapshot
ID, and can be obtained from 'emu log'. If omitted, all of the
snapshots within a sink will be verified. If no snapshot or sink is
specified, every sink will be verified.
.PP
Verification uses the same checksum algorithm which was used to
generate the snapshot ID. The results of verification are cached in
the snapshot node.
.SH OPTIONS
.TP
\-\-dont-cache
Disable the caching of verification results.
.TP
\-\-use-cached
Use cached values from a previous verification where available,
instead of computing new checksums.
.TP
\-S <dir> \-\-source-dir <dir>
Specify the directory to use as an emu source. Defaults to `.'.
.TP
\-v \-\-verbose
Increase verbosity.
.TP
\-\-version
Show version information and quit.
.TP
\-h \-\-help
Show this help and quit.
.SH EXAMPLES
.TP
Verify all snapshots for all sinks:
emu verify
.TP
Verify all snapshots for sink `origin':
emu verify origin
.TP
Verify a specific snapshot by ID:
emu verify origin:53b1e624ccedee864cbba76563522be43360ded5
.TP
Verify last three snapshots:
emu verify origin:HEAD origin:HEAD~ origin:HEAD~2
.SH EMU
Part of the
.B emu
(1)
suite
.SH AUTHOR
Chris Cummins <chrisc.101@gmail.com>
.SH SEE ALSO
.B emu
(1)


CHANGELOG
emu-0.0.1 (Thu Jan 17 11:51:54 GMT 2013)
--------------------------------------------------------------------------------

Date:   Thu Jan 17 11:51:32 2013 +0000
    Rename make-release mkrelease

Date:   Thu Jan 17 11:50:12 2013 +0000
    make: remove release rule

Date:   Thu Jan 17 11:47:45 2013 +0000
    Add release documentation

Date:   Thu Jan 17 11:32:08 2013 +0000
    libemu: exit code vars

Date:   Thu Jan 17 11:27:21 2013 +0000
    libemu: update documentation

Date:   Thu Jan 17 11:15:07 2013 +0000
    make: Don't use $(QUIET) without good cause

Date:   Thu Jan 17 11:12:20 2013 +0000
    make: use '-' to separate words in rules

Date:   Thu Jan 17 11:10:30 2013 +0000
    test: Fix libemu var name

Date:   Mon Dec 17 16:11:32 2012 +0000
    make: Fix typo

Date:   Mon Dec 17 16:11:04 2012 +0000
    make: Rename 'default' to 'help'

Date:   Mon Dec 17 16:10:23 2012 +0000
    make: Remove 'VERSION' var

Date:   Mon Dec 17 16:09:37 2012 +0000
    make: Use portability variables

Date:   Mon Dec 17 16:07:16 2012 +0000
    make: Assign phony rules

Date:   Mon Dec 17 16:06:33 2012 +0000
    Use conventional $(QUIET) Makefile assignment

    - make V=1 will now be verbose

Date:   Thu Dec 6 20:56:11 2012 +0000
    libemu: Refactor command_not_found

Date:   Thu Dec 6 20:54:59 2012 +0000
    Add global constant $EMU_VERSION

Date:   Thu Dec 6 20:53:40 2012 +0000
    libemu: Correct echo usage

Date:   Thu Dec 6 20:50:10 2012 +0000
    test: Add libemu echo test

Date:   Thu Dec 6 20:42:05 2012 +0000
    Remove $PROG variable

    - Refactor '$PROG' -> 'emu'
Merge: 2b4a93e 53ac6b2

Date:   Thu Dec 6 20:37:48 2012 +0000
    Merge branch 'wip/echo' into next

Date:   Thu Dec 6 20:36:49 2012 +0000
    libemu: Standardise echo functions

    - There are three functions providing message printing for emu:
      emu_echo(), emu_warning() and emu_error().

Date:   Thu Dec 6 20:08:54 2012 +0000
    Tidy use of emu_echo and emu_error

    - Messages begin with lowercase letter.

    - Message do not end in a full stop or other punctuation.

Date:   Thu Dec 6 20:05:54 2012 +0000
    emu: Fix error output

Date:   Thu Dec 6 20:05:27 2012 +0000
    Add EMU_ECHO_PREFIX constant

Date:   Thu Dec 6 17:06:26 2012 +0000
    libemu: Add emu_echo() and emu_error() functions

Date:   Thu Dec 6 17:06:12 2012 +0000
    Fix makefile lib target location

Date:   Thu Dec 6 14:12:38 2012 +0000
    libemu: Refactor with 'local' variables

Date:   Thu Dec 6 14:10:05 2012 +0000
    libemu: Adds exit_command_not_found()

Date:   Fri Nov 30 15:59:16 2012 +0000
    libemu: Add emu_echo() and emu_error()

Date:   Mon Nov 26 22:54:10 2012 +0000
    Fix Makefile rules

Date:   Sun Nov 25 21:38:18 2012 +0000
    Add make-release

Date:   Sun Nov 25 20:34:52 2012 +0000
    sink: Update man page

Date:   Sun Nov 25 20:34:43 2012 +0000
    libemu: Add unit tests

Date:   Sun Nov 25 20:34:38 2012 +0000
    sink: Add unit tests

Date:   Sun Nov 25 20:34:25 2012 +0000
    test: Add extra assertions

Date:   Sun Nov 25 20:33:57 2012 +0000
    libemu: Fix permissions test

Date:   Sun Nov 25 20:33:28 2012 +0000
    sink: Fix sinks list

Date:   Sun Nov 25 20:33:12 2012 +0000
    init: Fix command args

Date:   Sun Nov 25 16:25:19 2012 +0000
    libemu: Fix print_sink_details()

Date:   Sun Nov 25 15:17:44 2012 +0000
    sink: Add relative path support

Date:   Sun Nov 25 13:17:27 2012 +0000
    test: Use subdirectories for unit tests

Date:   Sun Nov 25 13:10:29 2012 +0000
    libemu: Add exit_if_no_dir_permissions()

Date:   Sun Nov 25 13:05:11 2012 +0000
    sink: Add man page

Date:   Sun Nov 25 12:51:14 2012 +0000
    emu: Add hook man page description

Date:   Sun Nov 25 12:51:05 2012 +0000
    reset: Fix man page whitespace

Date:   Sun Nov 25 12:50:54 2012 +0000
    restore: Add man page

Date:   Sun Nov 25 03:05:44 2012 +0000
    log: Fix man page

Date:   Sun Nov 25 03:05:32 2012 +0000
    reset: Add man page

Date:   Sun Nov 25 02:58:03 2012 +0000
    log: Add man page

Date:   Sun Nov 25 02:53:00 2012 +0000
    snapshot: Add man page

Date:   Sun Nov 25 02:52:45 2012 +0000
    init: Fix man page

Date:   Sun Nov 25 02:52:21 2012 +0000
    emu: Use man page

Date:   Sun Nov 25 02:21:06 2012 +0000
    init: Use man page documentation

Date:   Sun Nov 25 02:20:42 2012 +0000
    init: Use ENV variable

Date:   Sun Nov 25 01:57:25 2012 +0000
    Change /usr/share to /usr/local/share

Date:   Sun Nov 25 01:51:11 2012 +0000
    init: Add permissions check/test

Date:   Sun Nov 25 01:49:59 2012 +0000
    Add Makefile doc rule

Date:   Sun Nov 25 01:48:36 2012 +0000
    init: Implement verbosity

Date:   Sun Nov 25 00:55:13 2012 +0000
    Changed Makefile verbosity

Date:   Sun Nov 25 00:37:55 2012 +0000
    init: Implement unit tests

Date:   Sun Nov 25 00:36:05 2012 +0000
    init: Implement command args

Date:   Sat Nov 24 21:40:34 2012 +0000
    Add test-suite and Makefile rule

    - '$ make check' executes the test suite.

    - Tests are individual executables in the test/ directory.

    - Tests should be as verbose as possible, with a 'running-commentary' of
      progress printed to stdout and errors printed to stderr.

    - If a test fails, it should exit as soon as possible ('set -e' in bash
      scripts).

    - test/run-tests.sh executes the tests and prints a summary of the
      results.

Date:   Sat Nov 24 17:24:21 2012 +0000
    Change directory layout

    - emu-util has been renamed libemu
    - use /usr/local/lib for libemu

Date:   Sat Nov 24 17:09:07 2012 +0000
    restore: fixed stdout

Date:   Sat Nov 24 16:50:02 2012 +0000
    snapshot: just some fixes

Date:   Sat Nov 24 16:49:40 2012 +0000
    restore: fix help

Date:   Sat Nov 24 16:49:29 2012 +0000
    emu: add 'clean' description

Date:   Sat Nov 24 16:33:04 2012 +0000
    snapshot: Change stdout

Date:   Sat Nov 24 16:32:53 2012 +0000
    reset: First implementation

Date:   Sat Nov 24 16:32:42 2012 +0000

emu/emu-verify.py
#!/usr/bin/env python2.7
#
# Copyright 2015 Chris Cummins.
#
# Permission is hereby granted, free of charge, to any person
# obtaining a copy of this software and associated documentation files
# (the "Software"), to deal in the Software without restriction,
# including without limitation the rights to use, copy, modify, merge,
# publish, distribute, sublicense, and/or sell copies of the Software,
# and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

import re
import sys
import os

import emu as libemu
from libemu import EmuParser, Source, Colours, Util


def main(argv, argc):
    parser = EmuParser()
    parser.add_option("--dont-cache", action="store_true", dest="dont_cache",
                      default=False)
    parser.add_option("--use-cached", action="store_true", dest="use_cached",
                      default=False)
    (options, args) = parser.parse_args()

    snapshots = parser.parse_snapshots(Source(options.source_dir))
    status = 0

    if not len(snapshots):
        print "No snapshots to verify!"
        status = 1

    for snapshot in snapshots:
        if snapshot.verify(cache_results=not options.dont_cache,
                           use_cached=options.use_cached):
            print snapshot, Util.colourise("ok", Colours.OK)
        else:
            status = 2
            print snapshot, Util.colourise("bad", Colours.ERROR)

    return status


if __name__ == "__main__":
    argv = sys.argv[1:]
    ret = main(argv, len(argv))
    sys.exit(ret)
